CMakeLists.txt

```diff
@@ -0,0 +1,328 @@
+            set(SUBTARGETS Yes)
+        endif()
+    endforeach()
+    if(NOT SUBTARGETS)
+        message(WARNING "BUILD_GROUP is not defined and no Plugins were selected. Only the main binary will be built.")
+    endif()
+endif()
+
+if(RT_SLAM_PLUGINS)
+    set(USE_OFFLINE_CAM ON)
+    set(USE_OFFLINE_IMU ON)
+    set(USE_KIMERA_VIO ON)
+    set(USE_GTSAM_INTEGRATOR ON)
+    set(USE_POSE_PREDICTION ON)
+endif()
+
+if(CORE_PLUGINS)
+    set(USE_TIMEWARP_GL ON)
+    set(USE_AUDIO_PIPELINE ON)
+endif()
+
+find_package(PkgConfig REQUIRED)
+
+if("${CMAKE_SOURCE_DIR}" STREQUAL "${CMAKE_BINARY_DIR}")
+    message(FATAL_ERROR "In-source build prohibited.")
+endif("${CMAKE_SOURCE_DIR}" STREQUAL "${CMAKE_BINARY_DIR}")
+
+set(CMAKE_MODULE_PATH "${CMAKE_MODULE_PATH};${PROJECT_SOURCE_DIR}/cmake")
+
+#set(CMAKE_RUNTIME_OUTPUT_DIRECTORY runtime)
+
+add_definitions(-DILLIXR_BUILD_SUFFIX=${ILLIXR_BUILD_SUFFIX})
+add_executable(main${ILLIXR_BUILD_SUFFIX}.exe runtime/main.cpp)
+
+##############
+# dependencies
+##############
+
+##############
+# yaml parser
+##############
+
+include(cmake/GetYamlcpp.cmake)
+
+##############
+# opencv
+#
+# If building on centos system, opencv must be built from scratch to get the viz module
+# which is not supplied by the centos repos
+##############
+set(BUILDING_OPENCV No)
+set(OpenCV_DEP_STR "")
+if(HAVE_CENTOS)
+    set(INTERNAL_OPENCV ${CMAKE_INSTALL_PREFIX}/lib/cmake/opencv4)
+    # see if it has been built yet
+    find_package(OpenCV QUIET PATHS ${INTERNAL_OPENCV} NO_DEFAULT_PATH NO_CMAKE_ENVIRONMENT_PATH NO_SYSTEM_ENVIRONMENT_PATH NO_CMAKE_PACKAGE_REGISTRY NO_CMAKE_SYSTEM_PATH NO_CMAKE_SYSTEM_PACKAGE_REGISTRY)
+    if(NOT OpenCV_FOUND OR NOT OPENCV_VIZ_FOUND)
+	    message("Building OpenCV")
+        include(cmake/GetOpenCV_Viz.cmake)
+        set(USE_V4_CODE Yes)
+        add_dependencies(main${ILLIXR_BUILD_SUFFIX}.exe OpenCV_Viz)
+        set(OpenCV_DEP_STR OpenCV_Viz)
+        set(INTERNAL_OPENCV ${OpenCV_DIR})
+        set(BUILD_OPENCV Yes)
+    endif()
+else()
+    find_package(OpenCV REQUIRED)# CONFIG)
+endif()
+
+if(DEFINED OpenCV_VERSION_MAJOR)
+    if (${OpenCV_VERSION_MAJOR} EQUAL 4)
+        set(USE_V4_CODE Yes)
+    else()
+        set(USE_V4_CODE No)
+    endif()
+endif()
+
+find_package(GLEW REQUIRED)
+
+pkg_check_modules(glu REQUIRED glu)
+
+find_package(SQLite3 REQUIRED)
+
+find_package(X11 REQUIRED)
+
+pkg_check_modules(gl REQUIRED gl)
+
+#############
+# plugin specific dependencies
+#############
+
+if(USE_GTSAM_INTEGRATOR OR USE_ORB_SLAM OR USE_KIMERA_VIO OR USE_AUDIO_PIPELINE OR USE_MONADO OR USE_OPENXR_APP)
+    find_package(Git REQUIRED)
+endif()
+
+if (BUILD_GROUP STREQUAL "TESTS")
+    find_package(GTest REQUIRED)
+endif()
+
+if(USE_OFFLOAD_DATA)
+    find_package(Boost REQUIRED COMPONENTS filesystem)
+elseif(USE_ORB_SLAM)
+    find_package(Boost REQUIRED COMPONENTS filesystem serialization)
+endif()
+
+if(USE_GTSAM_INTEGRATOR OR USE_ORB_SLAM)
+    find_package(Eigen3 REQUIRED)
```

Comment:
I think more plugins use Eigen3 - for instance, gldemo and timewarp.

Reply:
Yes, there are several more. Some that don't include it directly, so I missed them initially. I think that since so many plugins require Eigen3 I will just make it required no matter what.

Reply:
Regarding gstreamer, is this a newer requirement as I do not see anything in the master branch that references gstreamer.

Reply:
Yes, we previously asked folks to install it manually, but that proved to be a lot of trouble since different distros package different versions. It would be nice if this could be handled by the build system.

Reply:
Which branch and plugin is it being used by? As I am making my changes off of the main branch and there is no mention of gstreamer anywhere in there.

Reply:
It's used by the offload_vio plugins in offload-h265-clean, which we're hoping to merge in the next release. I don't know how much work it involves to compile gstreamer and link to it statically, but, if it's a lot, I believe we can hold off on that and provide instructions for manual installation.

Reply:
Can I ask why gstreamer is statically linked?

Reply:
From looking at the code in that branch, the gstreamer libraries are all dynamically linked, not statically.

Reply:
Adding in the gstreamer dependency should be very straight forward.